src/coreclr/jit/valuenum.cpp

```diff
@@ -11119,32 +11223,26 @@ void Compiler::fgValueNumberPhiDef(GenTreeLclVar* newSsaDef, BasicBlock* blk, bo
                 sameVNP.SetBoth(ValueNumStore::NoVN);
             }
         }
-    }
 
-    // We should have visited at least one phi arg in the loop above
-    assert(phiVNP.GetLiberal() != ValueNumStore::NoVN);
-    assert(phiVNP.GetConservative() != ValueNumStore::NoVN);
+        phiArgSsaNums.Push(phiArg->GetSsaNum());
+    }
 
-    ValueNumPair newSsaDefVNP;
+    LclSsaVarDsc* newSsaDefDsc = lvaGetDesc(newSsaDef)->GetPerSsaData(newSsaDef->GetSsaNum());
+    ValueNumPair  newSsaDefVNP = newSsaDefDsc->m_vnPair;
 
     if (sameVNP.BothDefined())
     {
         // If all the args of the phi had the same value(s, liberal and conservative), then there wasn't really
         // a reason to have the phi -- just pass on that value.
         newSsaDefVNP = sameVNP;
     }
-    else
+    else if (!isUpdate)
     {
         // They were not the same; we need to create a phi definition.
-        ValueNum lclNumVN = ValueNum(newSsaDef->GetLclNum());
-        ValueNum ssaNumVN = ValueNum(newSsaDef->GetSsaNum());
-
-        newSsaDefVNP = vnStore->VNPairForFunc(newSsaDef->TypeGet(), VNF_PhiDef, ValueNumPair(lclNumVN, lclNumVN),
-                                              ValueNumPair(ssaNumVN, ssaNumVN), phiVNP);
+        newSsaDefVNP.SetBoth(
+            vnStore->VNForPhiDef(newSsaDef->TypeGet(), newSsaDef->GetLclNum(), newSsaDef->GetSsaNum(), phiArgSsaNums));
     }
```

Comment:
The new phi definition does not actually have any new information in it -- it is still the same {lcl num, ssa num, [list of ssa args]} structure, even after the update. So we can skip it . The only cases where it would change is if we were now able to prove that a predecessor was unreachable -- but diffs don't seem bad, and this way saves us having to do a map lookup for these.

Reply:
Although, I guess no map lookup should be necessary -- we can just check if the count of SSA args is the same. That should make this no-diff. I'll make that change.